### PR TITLE
Receiver pays fee by default

### LINKED ISSUE #363: Receiver pays fee for lsp by default

### PR DESCRIPTION
**Pull request by TonyGiorgio**
Fixes #363 & #373

Still need to test with the mutiny-web

### CODE REVIEW

File: mutiny-core/src/node.rs [hunk 1]
```diff
@@ -569,6 +569,68 @@ impl Node {
...
+            // LSP requires an amount
+            let amount_sat = amount_sat
+                .ok_or(MutinyError::BadAmountError)
+                .and_then(|amount| {
+                    if amount == 0 {
+                        Err(MutinyError::BadAmountError)
+                    } else {
+                        Ok(amount)
+                    }
+                })?;
```
Comment (benthecarman) [inline mutiny-core/src/node.rs:584]: can simplify this as `amount_sat.filter(|a| a > 0).ok_or(MutinyError::BadAmountError)`

File: mutiny-core/src/node.rs [hunk 2]
```diff
@@ -569,6 +569,68 @@ impl Node {
...
+            (amount_sat, None)
+        };
+
+        let invoice = self
+            .create_internal_invoice(amount_sat, lsp_fee_sat, description, route_hints)
+            .await?;
+
+        if let Some(lsp) = self.lsp_client.clone() {
+            self.connect_peer(PubkeyConnectionInfo::new(&lsp.connection_string)?, None)
+                .await?;
```
Comment (benthecarman) [inline mutiny-core/src/node.rs:607]: should check that we aren't connected already, I think we do this at the nodemanager level, not the node level

File: mutiny-core/src/node.rs [hunk 3]
```diff
@@ -569,6 +569,68 @@ impl Node {
...
+                        Err(MutinyError::BadAmountError)
+                    } else {
+                        Ok(amount)
+                    }
+                })?;
+
+            // check the fee from the LSP
+            let lsp_fee = lsp
+                .get_lsp_fee_sat(FeeRequest {
+                    pubkey: self.pubkey.to_string(),
```
Comment (benthecarman) [inline mutiny-core/src/node.rs:589]: .to_hex() might be better here

File: mutiny-core/src/node.rs (hunk 2 above)
Comment (TonyGiorgio) [reply mutiny-core/src/node.rs:601]: Not, this calls `connect_peer_if_necessary` so it does that check

File: mutiny-core/src/lspclient.rs [hunk 4]
```diff
@@ -50,8 +50,20 @@ pub struct ProposalResponse {
...
+pub struct FeeRequest {
+    pub pubkey: String,
+    pub amount_msat: u64,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct FeeResponse {
+    pub fee_amount_msat: u64,
+}
+
```
Comment (benthecarman) [inline mutiny-core/src/lspclient.rs:63]: for these would be good to derive debug and clone too

Comment (TonyGiorgio) [reply mutiny-core/src/lspclient.rs:63]: If and when needed :)

File: mutiny-core/src/node.rs [hunk 5]
```diff
@@ -569,8 +569,65 @@ impl Node {
...
+
+    async fn create_internal_invoice(
+        &self,
+        amount_sat: Option<u64>,
+        fee_amount_sat: Option<u64>,
+        description: String,
+        route_hints: Option<Vec<PhantomRouteHints>>,
     ) -> Result<Invoice, MutinyError> {
         let amount_msat = amount_sat.map(|s| s * 1_000);
+        let fee_amount_msat = fee_amount_sat.map(|s| s * 1_000);
```
Comment (benthecarman) [inline mutiny-core/src/node.rs:630]: You have it dividing by 1000 when we get the fee from the lsp, just to multiply it by 1000 here, should just have `get_lsp_fee_sat` return in msats instead so we don't lose precision for no reason. Can then rename to `get_lsp_fee_msat`

**Review by benthecarman — CHANGES_REQUESTED** [review]

File: mutiny-core/src/node.rs (hunk 5 above)
Comment (TonyGiorgio) [reply mutiny-core/src/node.rs]: fixed.

**Review by benthecarman — APPROVED** [review]

**Pull request merged**